keps/0018-controller-overhaul.md

```diff
@@ -0,0 +1,121 @@
+
+## Motivation
+
+### Goals
+
+- better understandable and tested controller code
+- ensuring correct plan is executed in case KUDO controller is not running for some time
+- atomicity on plan execution level - do not allow to run plan when other plan is in progress
+- be able to see current plan execution status
+
+### Non-Goals
+
+- providing longer history of plan executions that just the last one run
+- have manual plan execution feature (there are several approaches to how to do this, I think it deserves a separate KEP)
+
+## Proposal
+
+The proposed design that will solve all the current known problems with controllers and CRDs is to get rid of PlanExecution CRD altogether. Plan status will basically become instance status (it’s the case right now as well anyway, we just need to replicate those values across two places). 
+
+### Instance CRD changes
+
+The instance CRD for instance with running deploy plan might then look as following.
+
+```yaml
+ Instance
+  spec:
+    operatorVersion: operator-123
+    parameters:
+      - REPLICAS: 4
+  status:
+    planStatus: # (for every plan defined in OV, there will be one field)
+      - deploy:
+          status: IN_PROGRESS
+          executionName: deploy-1478631057
+          lastFinishedRun: 1568642057
+          overview:
+            strategy: serial
+            phases:
+              - name: deploy
+                strategy: serial
+                state: IN_PROGRESS
+                steps:
+                  - name: everything
+                    state: IN_PROGRESS
+                    delete: false
+      - upgrade: null # (never run)
+    state: IN_PROGRESS
+    activePlan: deploy-1478631057 # (will be null if no plan is run right now)
+    lastAppliedInstanceSpec:
+      version: 123 # probably some version from metadata?
+      spec:
+        operatorVersion: operator-123
+        parameters:
+          - REPLICAS: 3
+```
+`planStatus` is a property that basically replaces the current PlanExecution CRD - it reports on the status of the plans that are run right now or last runs of all the plans available for that operator version. This is also what `kudo plan history` and `kudo plan status` will query to get overview of the plans.
+
+`lastAppliedInstanceSpec` is persisting the state of the instance from the previous successful deploy/upgrade/update plan finished. We need this to be able to solve flaw n.1 described in the summary. This gets updated after a plan succeeds.
+
+`status.state` is just a simple way how to query the overall state of the instance (if any plan is running on it or not). Could be used by user as well as e.g. admission webhook to figure out if any changes to the CRD are allowed.
+
+### Admission webhook
+
+Part of the solution (addressing problem n.3 from the summary) is an admission webhook. This one will guard the Instance CRD and will disallow any changes in a spec if plan is running.
```

Comment:
Is admission hook part of the refactoring (1), or does the refactoring allow us to add the hook later (2), solving consistency issues? If (2) is the case, then we should probably mention it as an optional goal, that can be tackled in a separate effort.

Reply:
Good point, I've marked it as "stretch goal" in the following paragraph